README.md
# grunt-image-resize

[![Build Status](https://travis-ci.org/excellenteasy/grunt-image-resize.svg)](https://travis-ci.org/excellenteasy/grunt-image-resize)
[![Dependency Status](https://david-dm.org/excellenteasy/grunt-image-resize.svg)](https://david-dm.org/excellenteasy/grunt-image-resize)
[![devDependency Status](https://david-dm.org/excellenteasy/grunt-image-resize/dev-status.svg)](https://david-dm.org/excellenteasy/grunt-image-resize#info=devDependencies)

> Resizing images made easy - thanks to ImageMagick.

## Getting Started
This plugin requires Grunt `^0.4.5` and ImageMagick.

### Grunt
If you haven't used [Grunt](http://gruntjs.com/) before, be sure to check out the [Getting Started](http://gruntjs.com/getting-started) guide, as it explains how to create a [Gruntfile](http://gruntjs.com/sample-gruntfile) as well as install and use Grunt plugins. Once you're familiar with that process, you may install this plugin with this command:

```shell
npm install grunt-image-resize --save-dev
```

One the plugin has been installed, it may be enabled inside your Gruntfile with this line of JavaScript:

```js
grunt.loadNpmTasks('grunt-image-resize');
```

### ImageMagick
Make sure ImageMagick is installed on your system and properly set up in your `PATH`.

Ubuntu:

```shell
$ apt-get install imagemagick
```

Mac OS X (using [Homebrew](http://brew.sh/)):

```shell
$ brew install imagemagick
```

Windows & others: 

[http://www.imagemagick.org/script/binary-releases.php](http://www.imagemagick.org/script/binary-releases.php)

Confirm that ImageMagick is properly set up by executing `convert -help` in a terminal.

## The "image_resize" task

### Overview
In your project's Gruntfile, add a section named `image_resize` to the data object passed into `grunt.initConfig()`.

```js
grunt.initConfig({
  image_resize: {
    options: {
      width: 100,
      height: 100,
      overwrite: true
    },
    your_target: {
      // Target-specific file lists and/or options go here.
    },
  },
})
```

### Options

#### options.width
Type: `Number`
Default value: `0` (only if height is defined)

A number value that is passed as pixel or percentage value to imagemagick.

#### options.height
Type: `Number`
Default value: `0` (only if width is defined)

A number value that is passed as pixel or percentage value to imagemagick.

#### options.overwrite
Type: `Boolean`
Default value: `true`

Determines whether file that already exist under this destination will be overwritten.

#### options.upscale
Type: `Boolean`
Default value: `false`

Determines whether images will be upscaled. If set to `false` (default), image will be copied instead of resized if it would be upscaled by resizing.

#### options.crop
Type: `Boolean`
Default value: `false`

Determines whether images will be cropped after resizing to exactly match `options.width` and `options.height`.

#### options.gravity
Type: `String`
Default value: `Center`
Possible values: `NorthWest`, `North`, `NorthEast`, `West`, `Center`, `East`, `SouthWest`, `South`, `SouthEast`

When cropping images this sets the image gravity. Doesn't have any effect, if `options.crop` is `false`.

#### options.concurrency
Type: `Number`
Default value: Number of CPUs

Determines how many resize operations are executed in parallel.

#### options.quality
Type: `Number`
Default value: `1`

Determines the output quality of the resized image. Ranges from `0` (really bad) to `1` (almost lossless). Only applies to jpg images.

#### options.autoOrient
Type: `Boolean`
Default value: `false`

Determines if resized image will be rotated according to EXIF information. Only applies to jpg images.

### Usage Examples

#### Default Options
In this example, the default options are used to resize an image to 100px width. So if the `test/fixtures/wikipedia.png` file has a width of 500px, the generated result would be a 100px wide `tmp/wikipedia.png`.

```js
grunt.initConfig({
  image_resize: {
    resize: {
      options: {
        width: 100
      },
      files: {
        'tmp/wikipedia.png': 'test/fixtures/wikipedia.png'
      }
    }
  }
})
```

#### Prevent overwriting
In this example, we prevent the destination file from being overwritten if it already exists. It the file `tmp/wikipedia.png` already exists, for example because we just ran the task configuration above, this would **not** overwrite `tmp/wikipedia.png`. The file `tmp/wikipedia.png` would still be 100px wide.

```js
grunt.initConfig({
  image_resize: {
    no_overwrite: {
      options: {
        width: 50,
        overwrite: false
      },
      files: {
        'tmp/wikipedia.png': 'test/fixtures/wikipedia.png'
      }
    }
  }
})
```

#### Allow upscaling
By default, the task does not resize images which would be upscaled. It only allows downscaling. You can allow upscaling by setting the `upscale` option to `true`. Otherwise images will copied instead of resized when they would be upscaled by resizing.

```js
grunt.initConfig({
  image_resize: {
    no_overwrite: {
      options: {
        width: 600,
        upscale: true
      },
      files: {
        'tmp/wikipedia.png': 'test/fixtures/wikipedia.png'
      }
    }
  }
})
```

#### Resize all the images in a directory

```js
grunt.initConfig({
  image_resize: {
    resize: {
      options: {
        width: 100,
      },
      src: 'src/*.JPG',
      dest: 'dest/'
    }
  }
})
```

## Contributing
In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Lint and test your code using [Grunt](http://gruntjs.com/).


CHANGELOG.md
<a name="1.0.0"></a>
## 1.0.0 (2015-01-18)


#### Features

* **task:** add option `autoOrient` ([e0c3b654](http://github.com/excellenteasy/grunt-image-resize/commit/e0c3b654148985bcdfe224b0bc132d01f01e6f0f))


<a name="0.4.0"></a>
## 0.4.0 (2014-09-05)




test/image_resize_test.js
'use strict';

var grunt = require('grunt'),
    gm    = require('gm').subClass({ imageMagick: true }),
    async = require('async'),
    fs    = require('fs');

gm.compare = gm.prototype.compare;

/*
  ======== A Handy Little Nodeunit Reference ========
  https://github.com/caolan/nodeunit

  Test methods:
    test.expect(numAssertions)
    test.done()
  Test assertions:
    test.ok(value, [message])
    test.equal(actual, expected, [message])
    test.notEqual(actual, expected, [message])
    test.deepEqual(actual, expected, [message])
    test.notDeepEqual(actual, expected, [message])
    test.strictEqual(actual, expected, [message])
    test.notStrictEqual(actual, expected, [message])
    test.throws(block, [error], [message])
    test.doesNotThrow(block, [error], [message])
    test.ifError(value)
*/

var createTest = function(test, filename) {
  return function(callback) {
    gm('tmp/'+filename).size(function(err, features) {
      gm('test/expected/'+filename).size(function(err, expected) {
        test.equal(features.width, expected.width);
        test.equal(features.height, expected.height);
        gm.compare('tmp/'+filename, 'test/expected/'+filename, 0.1, function (err, isEqual, difference) {
          if (err) {
            throw err;
          }
          test.ok(isEqual, 'tmp/'+filename+' is not equal to test/expected/'+filename+' (difference: '+difference+')');
          callback();
        });
        
      });
    });
  };
};

exports.image_resize = {
  setUp: function(done) {
    // setup here if necessary
    done();
  },
  resize: function(test) {
    test.expect(12);

    async.series([
      createTest(test, 'gnu.jpg'),
      createTest(test, 'Rhododendron.jpg'),
      createTest(test, 'wikipedia.png'),
      createTest(test, 'TeslaTurbine.png'),
    ], test.done);

  },
  upscale: function(test) {
    test.expect(9);

    async.series([
      createTest(test, "upscale.png"),
      createTest(test, "upscale2.png"),
      createTest(test, "no_upscale.png"),
    ], test.done);
  },
  crop: function(test) {
    test.expect(3);

    createTest(test, "crop.png")(test.done);
  },
  cropGravity: function(test) {
    test.expect(3);

    createTest(test, "crop_gravity.png")(test.done);
  },
  quality: function(test) {
    test.expect(4);

    async.series([
      createTest(test, 'quality.jpg'),
      function(callback) {
        fs.stat('tmp/quality.jpg', function(err, stats) {
          fs.stat('test/expected/quality.jpg', function(err, expected) {
            var epsilon = 1024;
            test.ok(expected.size - epsilon < stats.size && expected.size + epsilon > stats.size );
            callback();
          });
        });
      }
    ], test.done);
  },
  autoOrient: function(test) {
    test.expect(3);

    createTest(test, "rotate.jpg")(test.done);
  },
  all: function(test) {
    test.expect(30);

    async.series([
      createTest(test, "directory/crop.png"),
      createTest(test, "directory/gnu.jpg"),
      createTest(test, "directory/no_upscale.png"),
      createTest(test, "directory/quality.jpg"),
      createTest(test, "directory/Rhododendron.jpg"),
      createTest(test, "directory/TeslaTurbine.png"),
      createTest(test, "directory/upscale2.png"),
      createTest(test, "directory/upscale.png"),
      createTest(test, "directory/rotate.jpg"),
      createTest(test, "directory/wikipedia.png")
    ], test.done);
  }
};


tasks/image_resize.js
/*
 * grunt-image-resize
 * https://github.com/excellenteasy.com/grunt-image-resize
 *
 * Copyright (c) 2013 David Pfahler
 * Licensed under the MIT license.
 */

'use strict';

var gm    = require('gm').subClass({ imageMagick: true });
var async = require('async');
var path  = require('path');
var os    = require('os');

module.exports = function(grunt) {

  // Please see the Grunt documentation for more information regarding task
  // creation: http://gruntjs.com/creating-tasks

  grunt.registerMultiTask('image_resize', 'resize images made easy', function() {

    var done = this.async();
    var originalOptions = this.options();
    var options = this.options({
      overwrite: true,
      upscale: false,
      concurrency: os.cpus().length,
      crop: false,
      gravity: 'Center',
      quality: 1,
      autoOrient: false
    });
    var series = [];

    if (options.height == null && options.width == null) {
      return grunt.fail.warn("Neither height nor width defined.");
    }
    if (options.height == null && options.width) {
      options.height = 0;
    }
    if (options.width == null && options.height) {
      options.width = 0;
    }

    // Iterate over all specified file groups.
    this.files.forEach(function(list) {
      // Iterate over all files specified in the group
      list.src.forEach(function(f) {
        var dirname = (list.dest[list.dest.length-1] === '/') ? list.dest : path.dirname(list.dest),
            filepath = f, imOptions = {
          srcPath:  filepath,
          dstPath:  (list.dest[list.dest.length-1] !== '/') ? list.dest : path.join(dirname, path.basename(filepath)) ,
          width:    options.width,
          height:   options.height,
          quality:  options.quality,
          autoOrient: options.autoOrient
        };

        // Prevent failing if destination directory does not exist.
        if (!grunt.file.isDir(dirname)) {
          grunt.file.mkdir(dirname);
        }

        if (options.overwrite === false && grunt.file.isFile(imOptions.dstPath)) {
          return grunt.log.writeln("Skipping "+filepath+" because destination already exists.\n"+
            "Set options 'overwrite' to true to enable overwriting of files.");
        }

        series.push(function(callback) {
          // Fail when image would be upscaled unless explicitly allowed
          gm(filepath).size(function(err, size) {
            if (!options.upscale &&
              ((originalOptions.width && size.width < originalOptions.width) ||
              (originalOptions.height && size.height < originalOptions.height))) {
              grunt.log.writeln("Copying "+filepath+" instead of resizing, because image would be upscaled.\n"+
                "To allow upscaling, set option 'upscale' to true.");
              grunt.file.copy(filepath, imOptions.dstPath);
              grunt.log.ok("Image "+filepath+" copied to "+imOptions.dstPath);
              callback();
            }
            else {
              var resizer;
              if (options.crop) {
                resizer = gm(filepath)
                  .resize(imOptions.width, imOptions.height, "^")
                  .gravity(options.gravity)
                  .crop(imOptions.width, imOptions.height);
              } else {
                resizer = gm(filepath)
                  .resize(imOptions.width, imOptions.height);
              }

              if (options.autoOrient) {
                resizer.autoOrient();
              }

              resizer
                .quality(Math.floor(imOptions.quality * 100))
                .write(imOptions.dstPath, function(err) {
                if (err) {
                  grunt.fail.warn(err.message);
                } else {
                  grunt.log.ok('Image '+filepath+' resized to '+ list.dest);
                }
                return callback();
              });
            }
          });
        });
      });
    });

    async.parallelLimit(series, options.concurrency, done);
  });

};


Gruntfile.js
/*
 * grunt-image-resize
 * https://github.com/excellenteasy.com/grunt-image-resize
 *
 * Copyright (c) 2013 David Pfahler
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    jshint: {
      all: [
        'Gruntfile.js',
        'tasks/*.js',
        '<%= nodeunit.tests %>'
      ],
      options: {
        jshintrc: '.jshintrc'
      },
    },

    // Before generating any new files, remove any previously-created files.
    clean: {
      tests: ['tmp/*']
    },

    // Configuration to be run (and then tested).
    image_resize: {
      options: {
        width: 100
      },
      resize: {
        files: [
          {'tmp/gnu.jpg': 'test/fixtures/gnu.jpg'},
          {'tmp/wikipedia.png': 'test/fixtures/wikipedia.png'},
          {'tmp/Rhododendron.jpg': 'test/fixtures/Rhododendron.jpg'},
          {'tmp/TeslaTurbine.png': 'test/fixtures/TeslaTurbine.png'}
        ]
      },
      no_overwrite: {
        options: {
          width: 0,
          height: 50,
          overwrite: false
        },
        files: [
          {'tmp/gnu.jpg': 'test/fixtures/gnu.jpg'},
          {'tmp/wikipedia.png': 'test/fixtures/wikipedia.png'},
          {'tmp/Rhododendron.jpg': 'test/fixtures/Rhododendron.jpg'},
          {'tmp/TeslaTurbine.png': 'test/fixtures/TeslaTurbine.png'}
        ]
      },
      upscale: {
        options: {
          width: 600,
          height: 0,
          upscale: true
        },
        files: [
          {'tmp/upscale.png': 'test/fixtures/upscale.png'}
        ]
      },
      upscale2: {
        options: {
          width: 600,
          height: 600,
          upscale: true
        },
        files: [
          {'tmp/upscale2.png': 'test/fixtures/upscale2.png'}
        ]
      },
      no_upscale: {
        options: {
          width: 600,
          height: 0,
          upscale: false
        },
        files: [
          {'tmp/no_upscale.png': 'test/fixtures/no_upscale.png'}
        ]
      },
      crop: {
        options: {
          width: 400,
          height: 300,
          upscale: false,
          crop: true
        },
        files: [
          {'tmp/crop.png': 'test/fixtures/crop.png'}
        ]
      },
      cropGravity: {
        options: {
          width: 400,
          height: 300,
          upscale: false,
          crop: true,
          gravity: "NorthWest"
        },
        files: [
          {'tmp/crop_gravity.png': 'test/fixtures/crop.png'}
        ]
      },
      quality: {
        options: {
          width: 600,
          height: 0,
          upscale: false,
          quality: 0.2
        },
        files: [
          {'tmp/quality.jpg': 'test/fixtures/quality.jpg'}
        ]
      },
      autoOrient: {
        options: {
          autoOrient: true,
          width: 100,
          height: 100
        },
        files: [
          {'tmp/rotate.jpg': 'test/fixtures/rotate.jpg'}
        ]
      },
      all: {
        options: {
          width: 100,
        },
        src: 'test/fixtures/*.*',
        dest: 'tmp/directory/'
      }
    },

    // Unit tests.
    nodeunit: {
      tests: ['test/*_test.js']
    },

    release: {
      email: 'bot@excellenteasy.com',
      name: 'excellenteasybot'
    }
  });

  // Actually load this plugin's task(s).
  grunt.loadTasks('tasks');

  require('load-grunt-tasks')(grunt);

  // Whenever the "test" task is run, first clean the "tmp" dir, then run this
  // plugin's task(s), then test the result.
  grunt.registerTask('test', [
    'clean',
    'image_resize',
    'nodeunit'
  ]);

  // By default, lint and run all tests.
  grunt.registerTask('default', ['jshint', 'test']);
};
